internal/cmd/log-file-metric-exporter/main.go

```diff
@@ -0,0 +1,150 @@
+package main
+
+import (
+	"flag"
+	"github.com/ViaQ/logerr/log"
+	"os"
+	"path"
+	"strings"
+	"net/http"
+	"github.com/openshift/cluster-logging-operator/internal/pkg/symnotify"
+        "github.com/prometheus/client_golang/prometheus"
+        "github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+
+
+
+var (
+ verbosity int = 0
+ namespace string = "unknown"
+ podname string = "unknown"
+ containername string ="unknown"
+)
+
+
+
+type FileWatcher struct {
+        watcher *symnotify.Watcher
+        metrics *prometheus.CounterVec
+        sizes  map[string]float64
+        added  map[string]bool
+}
+
+
+func (w *FileWatcher) Update(path string, namespace string, podname string, containername string) error {
+        counter, err := w.metrics.GetMetricWithLabelValues(path,namespace,podname,containername)
+        if err != nil {
+                return err
+        }
+        stat, err := os.Stat(path)
+        if err != nil {
+                return err
+        }
+        if stat.IsDir() {
+                return nil // Ignore directories
+        }
+        lastSize, size := float64(w.sizes[path]), float64(stat.Size())
+        w.sizes[path] = size
+        var add float64
+        if size > lastSize {
+                // File has grown, add the difference to the counter.
+                add = size - lastSize
+        } else if size < lastSize {
+                // File truncated, starting over. Add the size.
+                add = size
+        }
+        log.V(2).Info("For logfile in...", "path",path,"lastsize",lastSize,"currentsize",size,"addedbytes",add)
+	counter.Add(add)
+        return nil
+}
+
+func (w *FileWatcher) Watch( ) {
+
+        
+	for {
+		//All logfiles with containername are added to the watcher
+		//write event for these logfiles are being watched
+		//create event gets issued for all new logfiles appear under logfilepathname /var/log/containers/
+		//For the cases new log files added, old files moved, old files deleted, you need to add/remove them from watcher as whole dir added to the watcher
+		//For new log files added write event is not getting issued
+		
+                e, err := w.watcher.Event()
+                if err != nil {
+		log.Error(err, "Watcher.Event returning err")
+		os.Exit(1)
+		}
+
+                log.V(2).Info("Events notified for...", "e.Name",e.Name,"Event",e.Op)
+
+		//Get namespace, podname, containername from e.Name - log file path
+		//filename := strings.Trim(e.Name,"/var/log/containers/")
+		_,filename := path.Split(e.Name)
+                filename = strings.TrimRight(filename,".log")
+                log.V(2).Info("file name after trimming","filename", filename)
+                filenameslice := strings.Split(filename,"_")
+		if (len(filenameslice) == 3) {
+		  podname = filenameslice[0]
+                  namespace = filenameslice[1]
+                  containername = filenameslice[2]
+	        } else {
+	          namespace="unknown"
```

Comment:
I don't see how this will ever be the case since the file format is a known pattern that has existed since the inception of kubernetes.  If we don't match this means:
* there is an error in the logic
* the spec changed and we didn't make adjustments
The only open question I have is container name which based on your analysis is there.  I'm hoping it is always there.

Reply:
@jcantrill do we have situations where workloads dumping some data to /var/log/containers/  directly using their own filename conventions ? @alanconway kindly guide me here..

Reply:
no. we should outright  do nothing for files that do not match the pattern.  the pattern is the API we have from kubernetes